client/service/src/storage_monitor.rs

```diff
@@ -0,0 +1,143 @@
+	threshold: u64,
+	/// keeps the ref for file system watcher
+	_watcher: RecommendedWatcher,
+}
+
+impl StorageMonitorService {
+	/// creates new StorageMonitorService for given client config
+	pub fn new_for_config(config: &Configuration) -> Result<Option<Self>, Error> {
+		Ok(match (config.available_storage_theshold, config.database.path()) {
+			(0, _) => {
+				log::info!(
+					target: LOG_TARGET,
+					"StorageMonitorService: threshold 0 given, storage monitoring disabled",
+				);
+				None
+			},
+			(_, None) => {
+				log::warn!(
+					target: LOG_TARGET,
+					"StorageMonitorService: no database path to observe",
+				);
+				None
+			},
+			(threshold, Some(path)) => {
+				let (sink, stream) = tracing_unbounded("mpsc_storage_monitor", 1024);
+
+				let mut watcher = RecommendedWatcher::new(
+					move |res| {
+						sink.unbounded_send(res).unwrap();
+					},
+					Config::default(),
+				)
+				.map_err(|e| Error::Other(format!("Could not create fs watcher {e}")))?;
+
+				watcher
+					.watch(path.as_ref(), RecursiveMode::Recursive)
+					.map_err(|e| Error::Other(format!("Could not start fs watcher {e}")))?;
+
+				log::debug!(
+					target: LOG_TARGET,
+					"Initializing StorageMonitorService for db path: {:?}",
+					path,
+				);
+
+				Self::check_free_space(&path, threshold)?;
+
+				Some(StorageMonitorService {
+					path: path.to_path_buf(),
+					stream,
+					threshold,
+					_watcher: watcher,
+				})
+			},
+		})
+	}
+
+	/// main monitoring loop, intended to be spawn as essential task
+	pub async fn run(mut self) {
+		while let Some(watch_event) = self.stream.next().await {
+			match watch_event {
+				Ok(_) =>
+					if Self::check_free_space(&self.path, self.threshold).is_err() {
+						break
+					},
+				Err(e) => {
+					println!("watch error: {:?}", e);
+				},
+			}
+		}
+	}
+
+	// returns free space in MB
+	fn free_space(path: &Path) -> Result<u64, Errno> {
+		let fs_stats = statvfs(path);
+		fs_stats.map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
+	}
+
+	// checks if the free space for given `path` is below given `threshold`.
+	// If not error is returned.
+	// System errors are silently ignored.
+	fn check_free_space(path: &Path, threshold: u64) -> Result<(), Error> {
+		match StorageMonitorService::free_space(path.as_ref()) {
+			Ok(available_space) => {
+				log::trace!(
+					target: LOG_TARGET,
+					"free:{:?} , threshold:{:?}",
+					available_space,
+					threshold,
+				);
+
+				if available_space < threshold {
+					let msg = format!(
+								"Available space {:?}MB for path {:?} dropped below threshold:{:?}MB , terminating...",
+								available_space,
+								path,
+								threshold,
+							);
+					log::error!(target: LOG_TARGET, "{}", msg);
+					Err(Error::Other(msg))
+				} else {
+					Ok(())
+				}
+			},
+			Err(e) => {
+				log::warn!(target: LOG_TARGET, "could not read available space: {:?}", e);
```

Comment:
I don't see reason why `statvfs` would fail in normal conditions. But if that happens this warning may flood the console (~1Hz). Could be silenced for some period of time.
The other approach would be failing the essential task when `statvfs` fails, but it does not sound reasonable for me.
Any thoughts?

Reply:
One option would be omitting duplicate warning messages (if the error code has not changed between failures). But I don't know if storing and comparing the error codes worth it.

Reply:
What about a bool flag `statvfs_warn_logged`, which ensures only one warning is logged?